clib/memprof_coq.std.ml

```diff
@@ -0,0 +1,144 @@
+
+end
+
+module Thread_map_core = struct
+  open Resource_bind
+
+  module IMap = Map.Make (
+    struct
+      type t = int
+      let compare = Stdlib.compare
+    end)
+
+  type 'a t = { mutex : Mutex.t ; mutable map : 'a IMap.t }
+
+  let create () = { mutex = Mutex.create () ; map = IMap.empty }
+
+  let current_thread () = Thread.id (Thread.self ())
+
+  let get s =
+    (* Concurrent threads do not alter the value for the current
+       thread, so we do not need a lock. *)
+    IMap.find_opt (current_thread ()) s.map
+
+  (* For set and clear we need a lock *)
+
+  let set s v =
+    let& () = Mutex_aux.with_lock s.mutex in
+    let new_map = match v with
+      | None -> IMap.remove (current_thread ()) s.map
+      | Some v -> IMap.add (current_thread ()) v s.map
+    in
+    s.map <- new_map
+
+  let _clear s =
+    let& () = Mutex_aux.with_lock s.mutex in
+    s.map <- IMap.empty
+end
+
+module Masking = struct
+
+  module T = Thread_map_core
+
+  type mask = { mutable on : bool }
+
+  let mask_tls : mask T.t = T.create ()
+  (* whether the current thread is masked *)
+
+  let create_mask () =
+    let r = { on = false } in
+    T.set mask_tls (Some r) ;
+    r
+
+  let delete_mask () = T.set mask_tls None
+
+  let is_blocked () =
+    match T.get mask_tls with
+    | None -> false
+    | Some r -> r.on
+
+  let assert_blocked () = assert (is_blocked ())
+
+  (* The current goal is only to protect from those asynchronous
+     exceptions raised after dutifully checking that [is_blocked ()]
+     evaluates to false, and that expect the asynchronous callback to be
+     called again shortly thereafter (e.g. memprof callbacks). There is
+     currently no mechanism to delay asynchronous callbacks, so this
+     strategy cannot work for other kinds of asynchronous callbacks. *)
+  let with_resource ~acquire arg ~scope ~(release : _ -> unit) =
+    let mask, delete_after = match T.get mask_tls with
+      | None -> create_mask (), true
+      | Some r -> r, false
+    in
+    let old_mask = mask.on in
+    let remove_mask () =
+      (* remove the mask flag from the TLS to avoid it growing
+         uncontrollably when there are lots of threads. *)
+      if delete_after then delete_mask () else mask.on <- old_mask
+    in
+    let release_and_unmask r x =
+      match release r with
+      | () -> remove_mask () ; x
+      | exception e -> remove_mask () ; raise e
+    in
+    mask.on <- true ;
+    let r = try acquire arg with
+      | e -> mask.on <- old_mask ; raise e
+    in
+    match
+      mask.on <- old_mask ;
+      scope r
+    with
+    | (* BEGIN ATOMIC *) y -> (
+        mask.on <- true ;
+        (* END ATOMIC *)
+        release_and_unmask r y
+      )
+    | (* BEGIN ATOMIC *) exception e -> (
+        mask.on <- true ;
+        (* END ATOMIC *)
+        match Printexc.get_raw_backtrace () with
+        | bt -> (
+            let e = release_and_unmask r e in
+            Printexc.raise_with_backtrace e bt
```

Comment:
Does having to do this mean a exninfo may be lost?

Reply:
I don't know, this is a question for OCaml folks (@gadmm @gasche )

My understanding is that exninfo is in general not reliable and we should not attach anything to an exn that cannot be lost. Backtrackes will be fine with this tho as we did integrate with OCaml's system quite some time ago.

Reply:
I thought it was a question for Rocq folks (@ppedrot), since I think the implementation of exninfo is quite peculiar.

Reply:
I thought this was a question for the people understanding the OCaml runtime, so I guess we have reached a deadlock in the metatheory.

Reply:
I guess as long as the `scope` is low level / internal stuff like state resets the exninfo is always trivial so it doesn't matter, is that always the case?

Reply:
Through a quick peek it seems that the Exninfo is stored in some TLS, compared with physical equality. So I expect that reraising the same exception preserves Exninfo. But this seems to be entirely a matter of Rocq's implementation, so I will let the experts check that is correct.

Reply:
Actually there are stm bugs (number 5000) due to Coq losing exninfos, anyways:
- already exninfos are not very reliable, and we attach to them mostly info that is not critical to lose
- the current implementation should preserve the exninfo
- the masking happens only in very limited places, which as @SkySkimmer note, should attach anything important, for now the important cases are `Pcoq.unfreeze` and `dynload`, tho I think it would be convenient to protect all the summary restores (if there is not performance impact, which should not) , even if a bit more heavy-handed, it will interact with plugins much better.

So IMHO not a problem in this context.